From aab8b481570b78292de9ddc04c6b9392f72d8498 Mon Sep 17 00:00:00 2001
From: Alvaro Karsz <alvaro.karsz@solid-run.com>
Date: Tue, 18 Jan 2022 15:57:30 +0200
Subject: [PATCH] Read rootfs From SD card.

By default:
	booti won't pass rootfs address and bootargs points kernel to
	look for rootfs in SD card, partition 2.

User can run "run load_rootfs" in U-boot shell to load rootfs into RAM.

Signed-off-by: Alvaro Karsz <alvaro.karsz@solid-run.com>
---
 .../solidrun/am64x_solidrun/am64x_solidrun.c  | 23 +++++++++++++++++++
 configs/am64x_a53_solidrun_defconfig          |  2 +-
 include/configs/am64x_solidrun.h              | 15 ++++--------
 3 files changed, 29 insertions(+), 11 deletions(-)

diff --git a/board/solidrun/am64x_solidrun/am64x_solidrun.c b/board/solidrun/am64x_solidrun/am64x_solidrun.c
index bbb6a49c..b72b9034 100644
--- a/board/solidrun/am64x_solidrun/am64x_solidrun.c
+++ b/board/solidrun/am64x_solidrun/am64x_solidrun.c
@@ -8,6 +8,7 @@
 */
 
 #include <common.h>
+#include <command.h>
 #include <asm/io.h>
 #include <env.h>
 #include <spl.h>
@@ -666,3 +667,25 @@ void spl_board_init(void)
 	writel(val, AM64X_SR_CTRLMMR_USB0_PHY_CTRL);
 }
 #endif
+
+
+
+
+#ifndef	CONFIG_SPL_BUILD
+
+/*function to select relevant devicetree for HW configuration*/
+int find_devicetree(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
+{
+	/* Currently, just one device tree exists, so no need to search.
+	 * This function will be expanded when more HW configuration will be available.
+	 */
+
+	env_set("fdtfile", "am642-solidrun.dtb");
+
+	return 0;
+}
+
+/*expose functions to uboot shell*/
+U_BOOT_CMD(find_dt, 1, 0, find_devicetree, "", "");
+
+#endif //CONFIG_SPL_BUILD
diff --git a/configs/am64x_a53_solidrun_defconfig b/configs/am64x_a53_solidrun_defconfig
index 6acc0e18..000932a1 100644
--- a/configs/am64x_a53_solidrun_defconfig
+++ b/configs/am64x_a53_solidrun_defconfig
@@ -26,7 +26,7 @@ CONFIG_DISTRO_DEFAULTS=y
 # CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
 CONFIG_SPL_LOAD_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x81000000
-CONFIG_BOOTCOMMAND="mmc dev 1;  load mmc 1:1 0x80480000 Image; load mmc 1:1 0x83000000 am642-solidrun.dtb; load mmc 1:1 0x83800000 rootfs.cpio;  booti 0x80480000 0x83800000 0x83000000;"
+CONFIG_BOOTCOMMAND="find_dt; run load_kern; run load_fdt; run set_boot_args; run run_kern;"
 CONFIG_BOARD_LATE_INIT=y
 CONFIG_SPL_SYS_MALLOC_SIMPLE=y
 CONFIG_SPL_STACK_R=y
diff --git a/include/configs/am64x_solidrun.h b/include/configs/am64x_solidrun.h
index 56b1c6d5..eacd53c1 100644
--- a/include/configs/am64x_solidrun.h
+++ b/include/configs/am64x_solidrun.h
@@ -60,18 +60,13 @@
 
 /* U-Boot general configuration */
 #define EXTRA_ENV_AM642_BOARD_SETTINGS					\
-	"findfdt="							\
-		"if test $board_name = am64x_gpevm; then " \
-			"setenv fdtfile k3-am642-evm.dtb; fi; " \
-		"if test $board_name = am64x_skevm; then " \
-			"setenv fdtfile k3-am642-sk.dtb; fi;" \
-		"if test $fdtfile = undefined; then " \
-			"echo WARNING: Could not determine device tree to use; fi; \0" \
 	"name_kern=Image\0"						\
 	"console=ttyS2,115200n8\0"					\
-	"args_all=setenv optargs earlycon=ns16550a,mmio32,0x02800000 "	\
-		"${mtdparts}\0"						\
-	"run_kern=booti ${loadaddr} ${rd_spec} ${fdtaddr}\0"
+	"load_kern=load mmc 1:1 ${kernel_addr_r} ${name_kern}\0"	\
+	"load_fdt=load mmc 1:1 ${fdt_addr_r} ${fdtfile}\0"	\
+	"load_rootfs=setenv rd_spec ${ramdisk_addr_r}; load mmc 1:1 ${rd_spec} rootfs.cpio\0"	\
+	"set_boot_args=setenv bootargs console=${console} root=/dev/mmcblk1p2 rw rootwait\0"	\
+	"run_kern=booti ${kernel_addr_r} ${rd_spec} ${fdt_addr_r}\0"
 
 /* U-Boot MMC-specific configuration */
 #define EXTRA_ENV_AM642_BOARD_SETTINGS_MMC				\
-- 
2.32.0

